### PR TITLE
Install PTF 

### PR DESCRIPTION
**Pull request by John-A-Davies**
This PR adds function to the Jenkins test, to APPLY the PTF after the FMID.  See #163 

It checks the Jenkins parameters; if an FMID is specified, it first installs that FMID, otherwise it installs the GA FMID already on Marist.  If a SYSMOD (PTF, USERMOD, APAR) is specified, it cleans the Zowe runtime directory and applies the SYSMOD.  Either way, it runs the smoke tests afterwards.

### CODE REVIEW

**Comment by John-A-Davies** [comment]
A set of new JCL job files is provided for the pipeline to install the SYSMOD using SMP/E.

File: scripts/alloc-APF-libs.sh [hunk 1]
```diff
@@ -0,0 +1,118 @@
+#!/bin/sh
```
Comment (stevenhorsman) [inline scripts/alloc-APF-libs.sh]: Is this script used by anything?

**Review by stevenhorsman — COMMENTED** [review]
Looks of commented out code and echos that could do with cleaning up
Is the JCL checked in custom jobs you need to drive the ptf process? How come this is so different from the base FMID install?

File: scripts/alloc-APF-libs.sh (hunk 1 above)
Comment (John-A-Davies) [reply scripts/alloc-APF-libs.sh]: It's not used in our repo.  It's a convenience script to create the datasets used by the XMEM server in its PROC.  The comment header should explain this.

**Comment by John-A-Davies** [comment]
I'll clean up extraneous echo statements and commented-out code.

Applying the PTF appears different from the FMID because the FMID carries the SMP/E jobs in a RELFILE, whereas the essential JCL that the PTF uses is presented in ZOWE.AZWE001.TMP0001.readme.htm for customers to read and tailor, so it's not in a form that's readily usable by the automation.  Therefore I created JCL for the PTF as a convenience to the customer, and to be consumable by automation.

File: scripts/alloc-APF-libs.sh (hunk 1 above)
Comment (stevenhorsman) [reply scripts/alloc-APF-libs.sh]: This script isn't shipped as part of the product, so if it isn't used in here it's useless

Comment (John-A-Davies) [reply scripts/alloc-APF-libs.sh]: Right.  I propose I move it to `zowe-install-packaging` in `scripts/utils`.

Comment (stevenhorsman) [reply scripts/alloc-APF-libs.sh]: Yes - let's remove it from this PR and can have a separate one later to add it there

Comment (John-A-Davies) [reply scripts/alloc-APF-libs.sh]: Done.

**Review by stevenhorsman — APPROVED** [review]

**Comment by jackjia-ibm** [comment]
I was looking at a SUCCESS build log of this PR: https://wash.zowe.org:8443/job/zowe-install-test/view/change-requests/job/PR-176/6/consoleFull and found few errors in the log. Will those errors matter?

```
12:23:36 install-SMPE-SYSMOD.sh INFO: JCL /ZOWE/tmp/Z1ALLOC.sed.jcl submitted
12:23:36 install-SMPE-SYSMOD.sh INFO: Checking for completion of jobname Z1ALLOC jobid 03691 after 2 seconds
12:23:36 install-SMPE-SYSMOD.sh ERROR: no return code for jobid 03691 PID=66146
12:23:36 install-SMPE-SYSMOD.sh DISPLAY JOB output was:
12:23:36 S0W1      2020080  12:****:45.43             ISF031I CONSOLE MYCONS ACTIVATED
12:23:36 S0W1      2020080  12:****:45.43            -$DJ03691,CC
12:23:36 S0W1      2020080  12:****:45.45  JOB03691   $HASP890 JOB(Z1ALLOC)   CC=(JCLERROR)
```

```
12:23:36 install-SMPE-SYSMOD.sh INFO: JCL /ZOWE/tmp/Z2ACCEPT.sed.jcl submitted
12:23:36 install-SMPE-SYSMOD.sh INFO: Checking for completion of jobname Z2ACCEPT jobid 03692 after 2 seconds
12:23:36 install-SMPE-SYSMOD.sh INFO: Checking for completion of jobname Z2ACCEPT jobid 03692 after 7 seconds
12:23:36 install-SMPE-SYSMOD.sh INFO: Checking for completion of jobname Z2ACCEPT jobid 03692 after 67 seconds
12:23:36 install-SMPE-SYSMOD.sh ERROR: job Z2ACCEPT(JOB03692) failed, RC=12
```

Looks like the first 2 steps (Z1ALLOC and Z2ACCEPT) are failed.

**Comment by OnnoVdT** [comment]
@jackjia-ibm 
The Z2ACCEPT failure is not a problem. It tries to accept the FMID, which was done earlier in job ZWE8ACPT as part of the FMID install. 

@John-A-Davies 
What is more troublesome are other messages in the log

```
12:23:36 S0W1      2020080  12:****:45.45  JOB03691   $HASP890 JOB(Z1ALLOC)   CC=(JCLERROR)
12:23:36 install-SMPE-SYSMOD.sh INFO: job Z1ALLOC(JOB03691) ended, RC=
```
So we can't allocate the datasets holding the new PTF. The JCL looked fine at first glance, with a minor remark that `//* Change ZOWEAD3 to the high level qualifier used to upload the ++SYSMOD.` goes past column 71, but luckily still less than 80. It does trigger comment line continuation, but luckily again, the next line is also a comment. You might want to rework your sed to strip JCL comment lines to avoid substitutions to expand a line too much.

The only explanation I see right now is that the datasets already exist, from a previous test run. I see you clean up USS files at the beginning of a run, we should do the same for datasets.

More troublesome is the elapss time of the copy step that follows
```
12:23:36 install-SMPE-SYSMOD.sh INFO: job Z1ALLOC(JOB03691) ended, RC=
12:23:36 
12:23:36 install-SMPE-SYSMOD.sh function runJob ended
12:23:36 script install-SMPE-SYSMOD.sh copying USS sysmod files to datasets
12:23:36 script install-SMPE-SYSMOD.sh copy complete
12:23:36 
12:23:36 install-SMPE-SYSMOD.sh function runJob started
```
That's the fastest copy of 500M I've seen on the Marist system, so I'm thinking the copy wasn't done at all.

Next the sysmod receive is successful. With us not being able to allocate datasets and a copy that did not happen, I'm pretty sure we just received the sysmods from a previous build, not the current one.

**Comment by jackjia-ibm** [comment]
Few suggestions not sure if they make sense:

- currently the script will continue if the job fail. A better way is for some steps, we should fail the script and exit with errors, some errors could be ok and we can continue, like Z2ACCEPT. Or Z2ACCEPT should check if FMID has been accepted and make the script exit if any step fails.
- add a validation step after all jobs are done. The step should make SMPE query and see if the PTF is installed.
- add validation before installation if we should apply this PTF, like FMID has been installed and version matches.
- remove datasets could be used during installation to avoid re-use last test dataset.

**Comment by stevenhorsman** [comment]
@OnnoVdT and @jackjia-ibm - I've fixed up the comments and copyright statements that were over 71 cols, but John is away for 2 weeks, so we might need some help from Jack and you with the other issues.
I think Jack's suggestions make sense and we maybe need to add some debug to investigate this sysmod copying issue if the data-set clean up doesn't help.
What's the best way to get these updates done? Jack - I think the clean up of the data sets is the most urgent - do we have any code around from the proclib clean up already that can do this?

**Comment by OnnoVdT** [comment]
The smpe shell scripts that deal with dataset allocation all have a cleanup embedded. For example, see `function _alloc` in `smpe/bld/smpe-service.sh`.
With John out, let me know if I need to try to squeeze in some development time in for this.

**Comment by stevenhorsman** [comment]
I've got `install-SMPE-SYSMOD.sh` to delete the sysmod data sets at the start of the script and done a listds that shows they are gone and `Z1ALLOC` is ending RC 0.
The copy echo statement is still taking 0s, but I have a hunch that is just due to the run-with-timeout logic in the build. I can see that my listds debug statements are printed out 4mins before, but the result all comes up with everything else in `install-SMPE-SYSMOD.sh` within the same second and the builds are still passing, so I think with the addition of the delete at the start and the tidying up of comments we might be ok.

**Comment by stevenhorsman** [comment]
Superceded by https://github.com/zowe/zowe-install-packaging/pull/1222

**Pull request closed without merging**